src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentMarkupEncodingPass.cs

```diff
@@ -65,6 +71,145 @@ public override void VisitHtml(HtmlContentIntermediateNode node)
                         }
                     }
                 }
+
+                // If we reach here, we don't have newlines, tabs or non-ascii characters in this node.
+                // If we can successfully decode all HTML entities(if any) in this node, we can safely let it call AddContent.
+                var decodedContent = new List<string>();
+                for (var i = 0; i < node.Children.Count; i++)
+                {
+                    var child = node.Children[i];
+                    if (!(child is IntermediateToken token) || !token.IsHtml || string.IsNullOrEmpty(token.Content))
+                    {
+                        // We only care about Html tokens.
+                        continue;
+                    }
+
+                    if (TryDecodeHtmlEntities(token.Content, out var decoded))
+                    {
+                        decodedContent.Add(decoded);
+                    }
+                    else
+                    {
+                        node.SetEncoded();
+                        return;
+                    }
+                }
+
+                // If we reach here, it means we have successfully decoded all content.
+                // Replace all token content with the decoded value.
+                for (var i = 0; i < node.Children.Count; i++)
```

Comment:
Decided to not combine the tokens because that will mean I'll need to create a new IR node type that accepts the content directly. Otherwise the CodeWriter will have to have be aware of this pass and assume only one token will be present.